### PR TITLE
[TNP] Fix python get element. partially fixes #15095

### LINKED ISSUE #15095: [TNP] TNP names still visible on app::PropertyLinkSub
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

It looks like even after fix #14561 TNP names still appear in python workbenches. (App:PropertyLinkSub still shows all the TNP name instead of just the base name)
image

see video how to reproduce:
https://github.com/FreeCAD/FreeCAD/pull/14561#issuecomment-2156134416

### Full version info

```shell
OS: Windows 11 build 22631
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.37735 +262 (Git)
Build type: Release
Branch: CAM-Simulator3
Hash: ce225649ac8c712d3b34e95d7d2d59bedd648da1
Python 3.8.10, Qt 5.15.2, Coin 4.0.1, Vtk 8.2.0, OCC 7.6.3
Locale: English/United States (en_US)
Installed mods: 
  * BIM_Workbench 2021.12.0
  * fasteners 0.5.24
  * Manipulator 1.5.7
  * OpenTheme 2024.6.22
  * PieMenu 1.5.1
  * Render 2024.1.23
  * SelectorToolbar
  * sheetmetal 0.4.21
  * TabBar (Disabled)
```

### Subproject(s) affected?

Core

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by shaise**
As discussed with @bgbsww [here:](https://github.com/FreeCAD/FreeCAD/issues/15095#issuecomment-2204279479)
The python shape function getElement does not handle TNP names. 
This is one of the reasons that causes issues with App::propertyLinkSub as mentioned in  #15095
When testing on LS3 this function indeed supports TNP long names.
The fix is basically using LS3 implementation.
A unit test function wad added as well.

### CODE REVIEW

File: src/Mod/PartDesign/PartDesignTests/TestTopologicalNamingProblem.py [hunk 1]
```diff
@@ -1455,6 +1455,27 @@ def testSubelementNames(self):
...
+        self.Doc.recompute()
+        if body.Shape.ElementMapVersion == "":  # Should be '4' as of Mar 2023.
+            return
+        map = pad.Shape.ElementMap
+        # Assert
+        for tnpName in map.keys():
+            element1 = body.Shape.getElement(tnpName)
+            element2 = body.Shape.getElement(map[tnpName])
+            self.assertTrue(element1.isSame(element2))
+
```
[the remark this answers is not in the document]
Comment (shaise) [reply src/Mod/PartDesign/PartDesignTests/TestTopologicalNamingProblem.py:1479]: Thanks done!

File: src/Mod/Part/App/TopoShapePy.xml [hunk 2]
```diff
@@ -813,8 +813,12 @@ infos contains additional info on the solutions. It is a list of tuples:
     </Methode>
     <Methode Name="getElement" Const="true">
       <Documentation>
-        <UserDocu>Returns a SubElement
-getElement(elementName) -> Face | Edge | Vertex
+        <UserDocu>
+Returns a SubElement
+getElement(elementName, [silent = False]) -> Face | Edge | Vertex
+elementName:  SubElement name - i.e. 'Edge1', 'Face3' etc. 
+              Accepts TNP names as well
```
[the remark this answers is not in the document]
Comment (shaise) [reply src/Mod/Part/App/TopoShapePy.xml]: Thanks done!

**Comment by Syres916** [comment]
Unfortunately at present IMHO this shouldn't be merged, while it fixes the `getElement` use in external workbenches, it doesn't fix the use in core Python workbenches such as CAM and possibly Fem.

At present when testing for a fix to #15223 the output should be `Face5` but it's `?Face5` which hopefully won't take too much fixing. More concerning is there's no CAM tests in _Drilling_, _Slot_, _PocketShape_ etc that have picked this up so I'll have a look at enhancing all those tests.

@FEA-eng would you mind please having a look at any Python file in Fem that contains the string `getElement(` and if it's looking up geometry feature `Vertex`, `Edge` or `Face` ensure that a change of the base geometry such as changing a constraint in a Sketch correctly finds the name of the geometry feature. If you get a failure could you possibly look at the corresponding test and let me know if you would like assistance improving it to catch these, TIA.

**Comment by Syres916** [comment]
The root cause of the `?Face5` output is occurring elsewhere in the code when the document is being Recomputed and the list of Base Geometry is being rebuilt. So I'll go ahead with creating the additional tests for CAM operations to be added in another PR.

Therefore this PR can be merged as it shouldn't be modified to just get around the question mark at the beginning of the Face name, the root cause needs addressing.

**Comment by shaise** [comment]
the issue of '?FaceXX` is completely parallel to getElement. There is no connection between the two, its just a missing functionality of getElement. 
when getElement gets a ? item it should return an error. This is normal behavior.

**Comment by bgbsww** [comment]
Shai is correct.  The '?' is the missing element marker that is added by the code when an earlier lookup fails - why this is happening is currently under research.

This doesn't solve the places where that happens, it just makes the python interface layer correct.  

Incidentally, that lookup fail and add of the missing marker ( "?") is happening in ::getSubObject which is overloaded and in PartFeature and SketchObject.  The PartFeature one appears to be where this is occurring.  Why, I still don't know yet.

**Comment by FEA-eng** [comment]
> @FEA-eng would you mind please having a look at any Python file in Fem that contains the string `getElement(` and if it's looking up geometry feature `Vertex`, `Edge` or `Face` ensure that a change of the base geometry such as changing a constraint in a Sketch correctly finds the name of the geometry feature. If you get a failure could you possibly look at the corresponding test and let me know if you would like assistance improving it to catch these, TIA.

There are a few occurrences but I'm not sure if all constraints rely on that: https://github.com/search?q=repo%3AFreeCAD%2FFreeCAD+%22getElement%22+path%3Asrc%2FMod%2FFem&type=code&p=1

@marioalexis84 I haven't thought about this before but does it mean that FEM needs some adjustments due to the TNP mitigation code? Can you help with that?

**Comment by bgbsww** [comment]
> 
> @marioalexis84 I haven't thought about this before but does it mean that FEM needs some adjustments due to the TNP mitigation code? Can you help with that?

Hopefully it doesn't, but tests to prove that it is working correctly is really the important challenge.

**Pull request merged**